### PR TITLE
[expo-updates] Support 204 status for no-op responses

### PR DESCRIPTION
**Pull request by wschurman**
# Why

https://github.com/expo/universe/pull/12258#issuecomment-1529977129

We need to support empty bodies for no-op responses. Semantically, a 204 makes the most sense for this response (no body content, no content-type). Therefore, we need to handle this case before trying to enter the methods that parse the manifest based on content-type.

# How

Handle 204 with empty body first when protocol is 1.

# Test Plan

Run new tests.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-updates/ios/EXUpdates/AppLoader/FileDownloader.swift [hunk 1]
```diff
@@ -123,6 +123,20 @@ internal final class FileDownloader: NSObject, URLSessionDataDelegate {
...
+          [url.absoluteString]
+        )
+        self.logger.error(message: errorMessage, code: UpdatesErrorCode.assetsFailedToLoad)
+        errorBlock(NSError(
+          domain: ErrorDomain,
+          code: FileDownloaderErrorCode.InvalidResponseError.rawValue,
+          userInfo: [NSLocalizedDescriptionKey: errorMessage]
+        ))
+        return
+      }
```
Comment (ide) [inline packages/expo-updates/ios/EXUpdates/AppLoader/FileDownloader.swift:138]: I'm not sure this is an error: if we get a 204, it's just an empty file (`touch empty.txt`)

File: packages/expo-updates/ios/EXUpdates/AppLoader/FileDownloader.swift [hunk 2]
```diff
@@ -336,12 +350,39 @@ internal final class FileDownloader: NSObject, URLSessionDataDelegate {
...
     let headerDictionary = httpResponse.allHeaderFields as! [String: Any]
+    let responseHeaderData = ResponseHeaderData(
+      protocolVersion: headerDictionary.optionalValue(forKey: "expo-protocol-version"),
+      serverDefinedHeadersRaw: headerDictionary.optionalValue(forKey: "expo-server-defined-headers"),
+      manifestFiltersRaw: headerDictionary.optionalValue(forKey: "expo-manifest-filters"),
+      manifestSignature: headerDictionary.optionalValue(forKey: "expo-manifest-signature")
+    )
+
+    guard let data = data else {
+      if responseHeaderData.protocolVersion == "1" && httpResponse.statusCode == 204 {
```
Comment (ide) [inline packages/expo-updates/ios/EXUpdates/AppLoader/FileDownloader.swift:367]: Similar comment as on the Kotlin code about the protocol version check.
Also, do we care if the status code is 204, or is a 200 OK as well? (But if there's an empty 4xx or 5xx, we want to fail.) Or does an empty 200 response create a Data object that is empty? I am trying to understand when `data` is nil/null vs an empty Data instance, if ever.

**Review by ide — CHANGES_REQUESTED** [review]

**Review by ide — APPROVED** [review]

**Review by douglowder — APPROVED** [review]

File: packages/expo-updates/ios/EXUpdates/AppLoader/Crypto.swift [hunk 3]
```diff
@@ -32,6 +32,11 @@ internal enum PEMType {
   }
 }
 
+private let ErrorDomain = "EXUpdatesCrypto"
+private enum CryptoErrorCode: Int {
+  case PublicKeyDownloadError = 1049
```
[the remark this answers is not in the document]
Comment (ide) [reply packages/expo-updates/ios/EXUpdates/AppLoader/Crypto.swift:37]: Note that in a comment. It helps to know that there explicitly was no intent (other than grepping for uniqueness).

File: packages/expo-updates/ios/EXUpdates/AppLoader/FileDownloader.swift (hunk 1 above)
Comment (ide) [reply packages/expo-updates/ios/EXUpdates/AppLoader/FileDownloader.swift]: Makes sense

File: packages/expo-updates/ios/EXUpdates/AppLoader/Crypto.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (wschurman) [reply packages/expo-updates/ios/EXUpdates/AppLoader/Crypto.swift:37]: We have many sets of these. I'll add a comment to all of them in a follow-up. https://linear.app/expo/issue/ENG-8502/add-comment-to-all-error-codes-in-ios-expo-updates-to-comment-on

**Pull request merged**